File: ios/SsdidWallet/Domain/MDoc/MsoVerifier.swift
import Foundation
import CryptoKit

/// Verifies MSO digests and validity windows for mdoc credentials.
enum MsoVerifier {

    /// Verify that an IssuerSignedItem's digest matches the expected value in the MSO.
    ///
    /// Per ISO 18013-5, the digest is computed over the tag-24-wrapped CBOR encoding
    /// of the IssuerSignedItem (i.e., the encoded CBOR data item tag wraps the item bytes).
    static func verifyDigest(
        item: IssuerSignedItem,
        mso: MobileSecurityObject,
        namespace: String
    ) -> Bool {
        guard let expectedDigest = mso.valueDigests[namespace]?[item.digestId] else {
            return false
        }

        guard let computedDigest = CborCodec.digestIssuerSignedItem(item, algorithm: mso.digestAlgorithm) else {
            return false
        }

        return computedDigest == expectedDigest
    }

    /// Verify that the current time falls within the MSO validity window.
    static func verifyValidity(_ mso: MobileSecurityObject) -> Bool {
        let now = Date()
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        // Try with fractional seconds first, then without
        let validFrom = formatter.date(from: mso.validityInfo.validFrom)
            ?? ISO8601DateFormatter().date(from: mso.validityInfo.validFrom)
        let validUntil = formatter.date(from: mso.validityInfo.validUntil)
            ?? ISO8601DateFormatter().date(from: mso.validityInfo.validUntil)

        guard let from = validFrom, let until = validUntil else {
            return false
        }

        return now >= from && now <= until
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/mdoc/MsoVerifier.kt
package my.ssdid.wallet.domain.mdoc

import com.upokecenter.cbor.CBORObject
import java.security.MessageDigest
import java.time.Instant
import java.time.ZonedDateTime
import java.time.format.DateTimeParseException

object MsoVerifier {

    fun verifyDigest(
        item: IssuerSignedItem,
        mso: MobileSecurityObject,
        namespace: String
    ): Boolean {
        val expectedDigest = mso.valueDigests[namespace]?.get(item.digestId)
            ?: return false

        // Per ISO 18013-5, the digest is computed over the tag-24-wrapped
        // CBOR encoding of the IssuerSignedItem
        val itemCbor = CBORObject.NewMap()
        itemCbor["digestID"] = CBORObject.FromObject(item.digestId)
        itemCbor["random"] = CBORObject.FromObject(item.random)
        itemCbor["elementIdentifier"] = CBORObject.FromObject(item.elementIdentifier)
        itemCbor["elementValue"] = CborCodec.toCborObject(item.elementValue)
        val itemBytes = itemCbor.EncodeToBytes()

        // Wrap in tag 24 (encoded CBOR data item) before hashing
        val tagged = CBORObject.FromObjectAndTag(itemBytes, 24)
        val taggedBytes = tagged.EncodeToBytes()

        val digest = when (mso.digestAlgorithm) {
            "SHA-256" -> MessageDigest.getInstance("SHA-256").digest(taggedBytes)
            "SHA-384" -> MessageDigest.getInstance("SHA-384").digest(taggedBytes)
            "SHA-512" -> MessageDigest.getInstance("SHA-512").digest(taggedBytes)
            else -> return false
        }

        return digest.contentEquals(expectedDigest)
    }

    fun verifyValidity(mso: MobileSecurityObject): Boolean {
        val now = Instant.now()
        return try {
            val validFrom = parseInstant(mso.validityInfo.validFrom)
            val validUntil = parseInstant(mso.validityInfo.validUntil)
            !now.isBefore(validFrom) && !now.isAfter(validUntil)
        } catch (_: Exception) {
            false
        }
    }

    private fun parseInstant(s: String): Instant =
        try { Instant.parse(s) }
        catch (_: DateTimeParseException) { ZonedDateTime.parse(s).toInstant() }
}
